spec/Staking.md

```diff
@@ -0,0 +1,340 @@
+
+state = {
+  stakeholders: [ "0x23..45", "0xAB..CD" ],
+  balance: {
+    "0xAB..CD": 790,
+  },
+  delegation: {
+    "0x23..45": [ ["0x01..23", 500] ],
+    "0xAB..CD": [ ["0x01..23", 100], ["0x23..45", 110] ]
+  },
+  revocations: [
+    ["0xAB..CD", "0x01..23", 124000, 50],
+    ["0x23..45", "0x01..23", 124100, 500],
+  ]
+}
+
+> after a block with timestamp 124000
+
+state = {
+  stakeholders: [ "0x23..45", "0xAB..CD" ],
+  balance: {
+    "0xAB..CD": 840,
+  },
+  delegation: {
+    "0x23..45": [ ["0x01..23", 500] ],
+    "0xAB..CD": [ ["0x01..23", 50], ["0x23..45", 110] ]
+  },
+  revocations: [
+    ["0x23..45", "0x01..23", 124100, 500],
+  ]
+}
+
+> after a block with timestamp 124100
+
+state = {
+  stakeholders: [ "0x23..45", "0xAB..CD" ],
+  balance: {
+    "0x23..45": 500,
+    "0xAB..CD": 840,
+  },
+  delegation: {
+    "0xAB..CD": [ ["0x01..23", 50], ["0x23..45", 110] ]
+  }
+}
+```
+
+# Fee distribution
+
+You pay fees to make a transaction, which is greater than a specified minimum fee for the transaction type. Transaction fees within a block are collected and CCS holders and the block author share this.
+
+CCS Holders share the sum of the minimum fees for the transaction. They get fees in proportion to the sum of the balance of CCS plus the total amounts of CCS that are delegated to someone.
+
+Fees distributed to CCS holders will be rounded down to an integer. The block author gets the rest, which is the sum of the remaining amount due to rounding and the amount that exceeds the minimum fee.
+
+## Example
+
+```
+state = {
+  stakeholders: [ "0x23..45", "0xAB..CD" ],
+  balance: {
+    "0x23..45": 450,
+    "0xAB..CD": 690,
+  },
+  delegation: {
+    "0x23..45": [ ["0x01..23", 50] ],
+    "0xAB..CD": [ ["0x01..23", 200], ["0x23..45", 110] ]
+  },
+  revocations: [
+    ["0xAB..CD", "0x01..23", 124000, 50],
+  }
+}
+```
+
+In a such state, an account "0x23..45" has a weight of 500 (balance 450 + delegation 50), and "0xAB..CD" has a weight of 1000 (balance 690 + delegation 200 + delegation 110, pending revocation will not affect to it). If a block has total fees of 110, and if the sum of the minimum fees for it is 35, each account will share the following amount of fees.
+
+```
+share of "0x23..45" = Math.floor(35 * ( 500 / (500 + 1000))) = 11
+share of "0xAB..CD" = Math.floor(35 * (1000 / (500 + 1000))) = 23
+remainng share (of author) = 110 - (11 + 23) = 76
+```
+
+# Revocation
+
+`BeginRevoke` will queue a pending revocation instead of not revoking a delegation immediately. The revocation will be delayed to a certain amount of time (`REVOKE_PENDDING_INTERVAL`) to prevent abuse. It will be processed when a block whose timestamp is greater or equal than `endTime` is created. The balance of `delegator` will be increased by `quantity` on the revocation, and the pending revocation will be popped out from the revocation queue. The queue and delegations that will become empty should be removed from the state.
```

Comment:
`BeginRevoke` will queue a pending revocation instead of revoking a delegation immediately. The revocation will be delayed by a certain amount of time (`REVOKE_PENDDING_INTERVAL`) to prevent abuse. It will be processed when a block whose timestamp is greater than or equal to the time when `endTime` is created. The balance of `delegator` will be increased by `quantity` upon revocation, and the pending revocation will be removed from the revocation queue. The queue and delegations that become empty should be removed from the state.